pageserver/src/tenant/timeline.rs

```diff
@@ -4235,87 +4237,147 @@ impl Timeline {
-                start = img_range.end;
-                let image_layer = image_layer_writer.finish(self, ctx).await?;
-                image_layers.push(image_layer);
+                if wrote_keys {
+                    // Normal path: we have written some data into the new image layer for this
+                    // partition, so flush it to disk.
+                    start = img_range.end;
+                    let image_layer = image_layer_writer.finish(self, ctx).await?;
+                    image_layers.push(image_layer);
+                } else {
+                    // Special case: the image layer may be empty if this is a sharded tenant and the
+                    // partition does not cover any keys owned by this shard.  In this case, to ensure
+                    // we don't leave gaps between image layers, leave `start` where it is, so that the next
+                    // layer we write will cover the key range that we just scanned.
+                    tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+                }
             } else {
-                // Special case: the image layer may be empty if this is a sharded tenant and the
-                // partition does not cover any keys owned by this shard.  In this case, to ensure
-                // we don't leave gaps between image layers, leave `start` where it is, so that the next
-                // layer we write will cover the key range that we just scanned.
-                tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+                // Metadata keys image layer creation.
+                let total_kb_reads_begin = ctx
+                    .vectored_access_delta_file_size_kb
+                    .load(AtomicOrdering::SeqCst);
+                let data = self
+                    .get_vectored_impl(
+                        partition.clone(),
+                        lsn,
+                        ValuesReconstructState::default(),
+                        ctx,
+                    )
+                    .await?;
+                let (data, total_kb_retrieved) = {
+                    let mut new_data = BTreeMap::new();
+                    let mut total_kb_retrieved = 0;
+                    for (k, v) in data {
+                        let v = v.map_err(CreateImageLayersError::PageReconstructError)?;
+                        total_kb_retrieved += KEY_SIZE + v.len();
+                        new_data.insert(k, v);
+                    }
+                    (new_data, total_kb_retrieved / 1024)
+                };
+                let total_kb_reads = ctx
+                    .vectored_access_delta_file_size_kb
+                    .load(AtomicOrdering::SeqCst)
+                    - total_kb_reads_begin;
+                // Data too small, or most data are garbage
+
+                let trigger_generation =
+                    total_kb_reads >= 16000 && total_kb_reads >= total_kb_retrieved * 10;
```

Comment:
I'm not a big fan on tracking how many bytes were read in each request. It's done for every request, but we don't need it every time as far as I understand. Could you use the number of layers visited instead?

Reply:
yep, I'll change it to num delta layers visited for now, and find better ways to model it. The reason why I used kbs instead of num of files is that it's hard to compute read amplification without the size information. But I think I can make a separate PR with benchmark/testing result for further discussion.